services/hanke-service/src/main/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusUpdateRequest.kt

```diff
@@ -301,6 +301,54 @@ data class InvoicingPostalAddressRequest(
     val city: String?,
 )
 
+@JvmName("hasChangesInKaivuilmoitusAlueList")
+fun List<KaivuilmoitusAlue>?.hasChanges(areas: List<KaivuilmoitusAlue>?): Boolean {
+    if (this == null) {
+        return areas != null
+    }
+    if (areas == null) {
+        return true
+    }
+    if (this.size != areas.size) {
+        return true
+    }
+    return this.any { it.hasChanges(areas.find { area -> area.hankealueId == it.hankealueId }) }
+}
+
+fun KaivuilmoitusAlue.hasChanges(area: KaivuilmoitusAlue?): Boolean {
```

Comment:
Could we make copies from the request and the existing areas without the tormaystarkastelutulos and compare those with the built-in compare?

Reply:
I fixed this by overriding equals and hashCode in Tyoalue so that they exclude tormaystarkasteluTulos. Then there is no need to change anything in hasChanges.

Reply:
Overriding the equals and hashCode makes the change global. This might be fine for now, since we don't compare them anywhere else or place them in sets. At some point, this might change. Also, if we add anything to TyoAlue, we need to remember to update the equals and hashCode methods.

The approach I suggested might be easier to maintain. It requires a just couple of changes and it's targeted at this specific problem.

In `KaivuilmoitusUpdateRequest.hasChanges`:
```kotlin
        val areas = areas?.map { it.withoutTormaystarkastelut() }
        val newAreas = applicationData.areas?.map { it.withoutTormaystarkastelut() }
        // ...
            areas != newAreas ||
```

In `KaivuilmoitusAlue`:
```kotlin
    fun withoutTormaystarkastelut() =
        copy(tyoalueet = tyoalueet.map { it.copy(tormaystarkasteluTulos = null) })
```

Reply:
Plus change HakemusServiceITest to not compare with isEqualTo directly but like this:
```
                        prop(KaivuilmoitusData::areas)
                            .isNotNull()
                            .single()
                            .transform { it.withoutTormaystarkastelut() }
                            .isEqualTo(area.withoutTormaystarkastelut())
```
Did the change.